### PR TITLE
feat(common): add graphite decorator

### PR DESCRIPTION
**Pull request by pismenskiy**

### CODE REVIEW

File: packages/common/src/main/ts/graphit-metric/graphite.service.ts [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    graphiteApiEndpoint,
+    // @ts-ignore
+    syncInterval,
+    // @ts-ignore
+    datacenter,
+    // @ts-ignore
+    environment,
+  } = {}) {
+    if (!GraphiteLogger.instance) {
+      GraphiteLogger.applicationName = applicationName || env.APP_NAME
```
Comment (antongolub) [inline packages/common/src/main/ts/graphit-metric/graphite.service.ts:38]: wtf?

File: packages/common/src/main/ts/graphit-metric/graphite.service.ts [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+  getMetricsCallbacks: Array<(args?: Array<any>) => any> = []
+
+  // @ts-ignore
+  private static instance: this
+  private static environment: string
+  private static datacenter: string
+  private interval: number | undefined;
+
+  constructor({
+    // @ts-ignore
```
Comment (oljekechoro) [inline packages/common/src/main/ts/graphit-metric/graphite.service.ts:24]: без не получится?

**Review by oljekechoro — CHANGES_REQUESTED** [review]

File: packages/common/src/main/ts/graphit-metric/gm-rpm.decorator.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import { constructDecorator } from '@qiwi/decorator-utils'
+
+import { GraphiteLogger } from './graphite.service'
+
+export const GmRpmDecorator = constructDecorator(
```
Comment (antongolub) [inline packages/common/src/main/ts/graphit-metric/gm-rpm.decorator.ts:5]: `@Metered`, not `@Rpm`

**Review by antongolub — CHANGES_REQUESTED** [review]

**Review by oljekechoro — APPROVED** [review]

File: packages/common/src/main/ts/graphit-metric/graphite.service.ts [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+  async getMetricsFromCallbacks() {
+    return this.getMetricsCallbacks.reduce(
+      async (acc, callback) => ({ ...acc, ...(await callback()) }),
+      {},
+    )
+  }
+
+  static enrichMetricName = (metricName: string) => {
+    return `$type.app.$cluster.${this.applicationName}.$host.${this.environment}-${this.datacenter}-${this.podId}.$metric.${metricName}`
```
Comment (antongolub) [inline packages/common/src/main/ts/graphit-metric/graphite.service.ts:92]: no

**Review by antongolub — CHANGES_REQUESTED** [review]

File: packages/common/tsconfig.json [hunk 5]
```diff
@@ -3,7 +3,7 @@
   "compilerOptions": {
     "rootDir": "./src/main/ts/",
     "baseUrl": "./src/main/ts/",
-    "tsBuildInfoFile": "./buildcache/.tsbuildinfo"
+    "tsBuildInfoFile": "./buildcache/.tsbuildinfo",
```
Comment (oljekechoro) [inline packages/common/tsconfig.json:6]: trailing slash в json?

File: packages/metric/src/main/ts/metric.service.ts [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    const data = await Promise.all(this.metricsCallbacks.map((el) => el()))
+    const flatData = data.reduce((acc, el) => ({ ...acc, ...el }), {})
+
+    return Object.entries(flatData).reduce((acc, [key, value]) => {
+      return { ...acc, ...{ [`${this.metricPrefix}.${key}`]: value } }
+    }, {})
+  }
+
+  private formatTimers() {
+    if (this.collectionTimer.lenght) return
```
Comment (oljekechoro) [inline packages/metric/src/main/ts/metric.service.ts:93]: length

File: packages/metric/src/main/ts/graphite.service.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+@Injectable()
+export class GraphiteService {
+  private client: GraphiteClient
+
+  constructor(graphiteApiEndpoint: string) {
+    this.client = GraphiteClient.createClient(`plaintext://${graphiteApiEndpoint}/`)
+  }
+
+  public sendMetric(metrics: Record<string, any>) {
+    return new Promise((resolve, reject) => {
```
Comment (oljekechoro) [inline packages/metric/src/main/ts/graphite.service.ts:13]: Promise<void>

**Review by oljekechoro — CHANGES_REQUESTED** [review]

**Review by oljekechoro — APPROVED** [review]

File: packages/metric/package.json [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    "@nestjs/common": "^9.2.1",
+    "@nestjs/core": "^9.2.1",
+    "@qiwi/decorator-utils": "^3.2.8",
+    "graphite": "^0.1.5",
+    "measured-reporting": "^2.0.0"
+  },
+  "devDependencies": {
+    "@qiwi/nestjs-enterprise-infra": "workspace:*",
+    "@types/graphite": "^0.1.0",
+    "esm": "^3.2.25",
```
Comment (antongolub) [inline packages/metric/package.json:51]: @qiwi/esm

File: packages/metric/package.json [hunk 9]
```diff
@@ -0,0 +1,66 @@
+{
+  "name": "@qiwi/nestjs-enterprise-metric",
+  "version": "1.6.1",
+  "description": "Logger module for nestjs enterprise",
+  "publishConfig": {
+    "access": "public"
+  },
+  "main": "./target/es6/index.js",
+  "exports": "./target/es6/index.js",
+  "source": "target/ts/index.ts",
```
Comment (antongolub) [inline packages/metric/package.json:10]: rm. This is for sourcemaps

File: packages/metric/package.json [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    "esm": "^3.2.25",
+    "reflect-metadata": "^0.1.13"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/qiwi/nestjs-enterprise.git"
+  },
+  "keywords": [
+    "js platform",
+    "logger"
```
Comment (antongolub) [inline packages/metric/package.json:60]: metrics

File: packages/metric/src/main/ts/decorators/error.decorator.ts [hunk 11]
```diff
@@ -0,0 +1,19 @@
+import { Inject } from '@nestjs/common'
+import { constructDecorator } from '@qiwi/decorator-utils'
+
+export const ErrorDecorator = constructDecorator(
+  ({ target, proto, args: [metricName] }) => {
+    const injectMetric = Inject('IMetricService')
+    injectMetric(proto, 'metricService')
+    return async function (...args: Array<any>) {
+      // @ts-ignore
+      return target.apply(this, args).catch((e: any) => {
```
Comment (antongolub) [inline packages/metric/src/main/ts/decorators/error.decorator.ts:10]: This may be sync

**Review by antongolub — CHANGES_REQUESTED** [review]

File: packages/metric/src/main/ts/decorators/error.decorator.ts [hunk 12]
```diff
@@ -5,15 +5,31 @@ export const ErrorDecorator = constructDecorator(
   ({ target, proto, args: [metricName] }) => {
     const injectMetric = Inject('IMetricService')
```
Comment (antongolub) [inline packages/metric/src/main/ts/decorators/error.decorator.ts:6]:
```js
const isPromise = (value) => value instanceof Promise || (typeof value?.then === 'function' && typeof value?.catch === 'faunction')
const onError = (e) => {...}
...

try {
  const res = target.apply(...)
  
  if (isPromise(res)) {
      return res.catch(onError)
  }
  
} catch (e) {
   onError(e)
}
```

**Review by antongolub — CHANGES_REQUESTED** [review]

**Review by antongolub — APPROVED** [review]

**Pull request merged**